plugin/formatting.py

```diff
@@ -56,11 +67,74 @@ def run(self, _):
                     "uri": filename_to_uri(self.view.file_name())
                 },
                 "range": Range.from_region(self.view, region).to_lsp(),
-                "options": {
-                    "tabSize": self.view.settings().get("tab_size", 4),
-                    "insertSpaces": True
-                }
+                "options": get_formatting_options(self.view)
             }
             client.send_request(Request.rangeFormatting(params),
                                 lambda response: self.view.run_command('lsp_apply_document_edit',
                                                                        {'changes': response}))
+
+
+class DocumentOnTypeFormattingListener(sublime_plugin.ViewEventListener):
+
+    @classmethod
+    def is_applicable(cls, subl_settings: sublime.Settings) -> bool:
+        syntax = subl_settings.get('syntax')
+        if not syntax:
+            return False
+        elif not is_supported_syntax(syntax):
+            return False
+        return True
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._triggers = []  # type: List[str]
+        self._client = None  # type: Optional[Client]
+        self._initialized = False
+
+    def on_modified_async(self) -> None:
+        if not settings.format_on_type:
+            return
+        if not self._ensure_initialized():
```

Comment:
For clarity I prefer what the CompletionHandler does:
```python
if not self._initialized:
    self._initialize()  # sets self._enabled

if self._enabled:
```

(in your implementation, `_ensure_initialized` means both "has been initialised" and "is enabled")

Reply:
done